# Lab | Inferential statistics - T-test & P-value

### Instructions

1. *One tailed t-test* - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file `files_for_lab/machine.txt`.
   Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

2. *Matched Pairs Test* - In this challenge we will compare dependent samples of data describing our Pokemon (file `files_for_lab/pokemon.csv`). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.


# OPTIONAL PART | Inferential statistics - ANOVA

Note: The following lab is divided in 2 sections which represent activities 3 and 4.

## Part 1

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on:
    - Null hypothesis
    - Alternate hypothesis
    - Level of significance
    - Test statistic
    - P-value
    - F table

### Context

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data `anova_lab_data.xlsx` file in the `files_for_lab` folder  

- State the null hypothesis
- State the alternate hypothesis
- What is the significance level
- What are the degrees of freedom of model, error terms, and total DoF

Data was collected randomly and provided to you in the table as shown: [link to the image - Data](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.05/7.05-lab_data.png)


## Part 2

- In this section, use the Python to conduct ANOVA.
- What conclusions can you draw from the experiment and why?

In [3]:
# Import Libraries
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

In [35]:
machine = pd.read_csv("/Users/robertkammerer/Ironhack/07_Week/day2_labs/lab-t-tests-p-values/files_for_lab/machine.csv")
machine.head()


New machine  Old machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5

In [40]:
print("pop mean new: ",machine['New machine'].mean())
print("pop mean old: ",round(machine['Old machine'].mean(),2))

pop mean new:  42.14
pop mean old:  43.23


##### State the Hypothesis

H0: new machine is equal or worse than the old machine $$H0: \mu ≥ 43.23$$


H1: new machine is better than the old machine $$H1: \mu < 43.23$$

##### Significance level 0.05

In [43]:
# not taking a sample since the dataset is super small anyway
m_sample = machine['New machine'].sample(5, random_state=1)
m_sample.head(5)

sample_mean = m_sample.mean()
sample_std = m_sample.std(ddof=1)
pop_mean = machine['New machine'].mean()
pop_std = machine['New machine'].std()
sem = sample_std/np.sqrt(30)
print("Population mean is: {:.2f}".format(pop_mean))
print("Population standard deviation is: {:.2f}".format(pop_std))
print()
print("Our sample mean is: {:.2f}".format(sample_mean))
print("Our sample standard deviation is: {:.2f}".format(sample_std))
print("The sem is: {:.2f}".format(sem))
print()
t = (sample_mean - pop_mean) / sem
print("Our statistic is: {:.2f}".format(t))
print("The p_value corresponding to our statistic is: {:.2f}".format(st.t.cdf(t,df = len(m_sample)-1)))
print("The significance level is set to 0.05")
print("We accept the H0?")

Population mean is: 42.14
Population standard deviation is: 0.68

Our sample mean is: 42.34
Our sample standard deviation is: 0.71
The sem is: 0.13

Our statistic is: 1.54
The p_value corresponding to our statistic is: 0.90
The significance level is set to 0.05
We accept the H0?


In [44]:
# Test would suggest to Accept the H0, which would be a error type I

### Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [8]:
pokemon = pd.read_csv("/Users/robertkammerer/Ironhack/07_Week/day2_labs/lab-t-tests-p-values/files_for_lab/pokemon.csv")
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

##### State the Hypothesis

$$H0: \mu_{atk} = \mu_{def} → \mu_{atk} - \mu_{def} = 0$$
$$H1: \mu_{atk} \neq \mu_{def} → \mu_{atk} - \mu_{def} \neq 0$$

In [50]:
#Create Sample

sample = pokemon[['Attack','Defense']].sample(30, random_state=3)
sample['difference'] = sample['Attack']-sample['Defense']
sample.head()

Attack  Defense  difference
108      30       50         -20
448      65       34          31
137     155      120          35
81       95      115         -20
144      55       50           5

In [51]:
sample_diff_mean, sample_diff_std = sample['difference'].mean(), sample['difference'].std(ddof=1)
sample_diff_mean, sample_diff_std

(-0.6666666666666666, 29.542877469796377)

In [52]:
t = sample_diff_mean / ( sample_diff_std / np.sqrt(sample.shape[0]) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: -0.67
The standard deviation of our samples differences is: 29.54
Our t statistics is: -0.12


In [53]:
tc = st.t.ppf(1-(0.05/2),df= sample.shape[0] - 1)
tc

2.045229642132703

In [54]:
st.t.cdf(t,df = sample.shape[0] - 1)

0.4512425543951809

In [56]:
st.ttest_rel(sample['Attack'],sample['Defense'])

Ttest_relResult(statistic=-0.12359946049379446, pvalue=0.9024851087903618)